[article] Vector Search in Production (#1609)
qdrant/landing_page: qdrant-landing/content/articles/vector-search-production.md

They had left HNSW indexing enabled, so every insert triggered a full index update. Unfortunately, their CPU usage soared, and other services timed out.

> ✅ On a case-by-case basis, we recommend **disabling building the HNSW index during large uploads** to improve ingestion and indexing speed. 

Once all records are inserted, you can rebuild the index in a single pass. Consider a specialized ingestion pipeline that batches writes and schedules indexing during low-traffic windows. If you don't have such low-traffic windows, you can tune the `indexing_threshold` to find a balance between receiving updates without triggering indexation, and keeping the collection indexed.

||
|-|
|**Read More:** [**Configuring the Vector Index**](https://qdrant.tech/documentation/concepts/indexing/#vector-index)|

### Other Solutions to Alleviate Indexing Bottleneck

✅ **Increase indexing threads:** If you're using more than 16 cores, consider explicitly increasing the number of indexing threads. 

> By default, Qdrant uses around 16 threads for indexing, but if you notice your CPU isn't being fully utilized during indexing, you can increase this number.

> ✅ **Optimize batch size and concurrency:** Keep your batch size at the default (around 100) and instead increase the number of concurrent processes. Running 50-60 concurrent processes can significantly improve upload performance. Using just one or two processes won't allow you to see the true performance potential.

Reviewer: Not so sure about this one. If you already use batching with wait=false it's likely that you already saturate upserting, in which case it will not be much faster with more parallel processes.

I think just stating 'use batching' is good enough.

**Be patient with indexing:** After uploading large datasets, there's a waiting period for indexing to complete. This is normal and can take time depending on your dataset size. 

||
|-|
|**Read More:** [**Configuration Documentation**](https://qdrant.tech/documentation/guides/configuration/)|

### When Indexing Falls Behind Ingestion
![vector-search-production](/articles_data/vector-search-production/vector-search-production-3.jpg)

> It's possible for indexing to temporarily fall behind data ingestion, both during gradual streaming uploads and after large bulk uploads.

> By default, searches include indexed data. However, a large number of unindexed points can significantly slow down searches due to full scans, potentially causing high search latency, timeouts, and application failures.

Reviewer: ```suggestion
By default, searches include unindexed data. However, a large number of unindexed points can significantly slow down searches due to full scans, potentially causing high search latency, timeouts, and application failures.
```

In this case, you should [**choose the right number of shards**](https://qdrant.tech/documentation/guides/distributed_deployment/#sharding) based on your node count and expected RPS.

You need to implement a shard strategy that aligns with real usage patterns. First, distribute your shards across all available nodes. This will help balance the load more effectively. After redistributing the shards, run performance tests to see how it affects your system. Then add replicas and test again to see how that changes performance.

Your sharding strategy also depends on how many collections you have. A single collection is arguably easier to balance, because there is less to move/balance. Also, a single collection also has the least amount of overhead/orchestration.

#### How to shard?
Proper sharding considers data distribution and query patterns. By default, shards are randomized for uniform distribution if queries always span the entire dataset. If certain tenants or geographical regions get hammered with traffic, you might partition by cluster, by payload (tenant ID) or a custom shard distribution.

||
|-|

> |**Read More:** [Sharding Documentation](https://qdrant.tech/documentation/concepts/sharding/)|

Reviewer: Consistency nitpick:

```suggestion
|**Read More:** [**Sharding Documentation**](https://qdrant.tech/documentation/concepts/sharding/)|
```

> If you have 3 nodes with just 1 shard, and replication factor 6. It will create 3 replicas (one on each node) of that shard, because it can't host more. If you add 3 more nodes at peak times, it'll automatically replicate that shard 3 more times in an attempt to match the factor of 6.

If you scale down again, the extra replicas will be dropped.

✅ **Reshard collections after scaling up** your cluster to rebalance data and avoid OOMs.

When you add nodes to your cluster, existing replicas rebalance themselves with new shards, but only with a fixed number - which may not fully take advantage of the new hardware. As a result, the original nodes remain overloaded while new nodes sit mostly idle. By **resharding** collections after scaling, you redistribute data evenly across the cluster, preventing hot spots that can lead to out-of-memory (OOM) conditions on overburdened nodes.

If new nodes remain empty after joining, you waste resources. If departing nodes leave behind un-migrated data, you risk partial coverage or even data loss.

> **Note:** This is only available in Qdrant Cloud as well as Hybrid & Private Clouds, and not when self hosting. 

||
|-|

> |**Read More:** [Distributed Deployment Documentation](https://qdrant.tech/documentation/guides/distributed_deployment/)|

Reviewer: ```suggestion
|**Read More:** [**Distributed Deployment Documentation**](https://qdrant.tech/documentation/guides/distributed_deployment/)|
|**Read More:** [**Resharding**](https://qdrant.tech/documentation/cloud/cluster-scaling/#resharding)|
```

Remember, cold-starts and query behaviour are dataset dependent, which is why you should establish your own baselines and what to expect.

||
|-|
|**Read More:** [Distributed Deployment Documentation](https://qdrant.tech/documentation/guides/distributed_deployment/)

### How to Design Your Systems to Protect Against Failure

||
|:-:|
|**"When one node crashed, our entire cluster went down."**|

Unfortunately, you didn't plan for failover or chaos testing, so a single point of failure took out production. You should plan for hardware or node crashes by storing data redundantly, testing failovers, and running chaos experiments.

✅ **Regularly test failures** to reveal how your system recovers.

High concurrency and large memory footprints can expose misconfigurations more quickly, so regularly simulating failures reveals how your system recovers.

- **Graceful Node Shutdowns**: Drain queries, reassign shard ownership via load balancer.
- **Redundant Data Paths**: Store data on multiple volumes or in multiple locations.
- **Load Tests**: Generate high concurrency or large queries to mimic real surge patterns.

### Set up Telemetry for Early Detection

> ❓ **Use Case:** A customer in health care uses telemetry data from their Qdrant deployment to identify performance and scaling issues with their open-source implementation. The telemetry helps them monitor metrics such as search performance, RAM utilization efficiency, and indexing speed. By analyzing this data, they can work toward reducing query response times from 50-60 seconds to 0.5 seconds and optimize their system configuration.

Reviewer 2: It sounds awful that someone can initially have 50-60 seconds of latency. Is that a real case? Also, 0.5 seconds is sometimes too much. I remember the last call with one of the users, and they reported less than 80ms.

![Grafa dashboard](/documentation/cloud/cloud-grafana-dashboard.png)

> Include tests that combine both read and write operations to simulate real-world usage. For example, you might configure a test with 80% reads and 20% writes to match your expected production workload.

By following these comprehensive load testing practices, you'll be able to identify and address potential bottlenecks before your system goes live, ensuring a smoother launch and better user experience.

||
|-|
|**Read More:** [**Telemetry and Monitoring Documentation**](https://qdrant.tech/documentation/guides/monitoring/)|
|**Read More:** [**Cloud Monitoring Documentation**](https://qdrant.tech/documentation/hybrid-cloud/networking-logging-monitoring/)

## 4. Ensuring Disaster Recovery With Database Backups and Snapshots
![vector-search-production](/articles_data/vector-search-production/vector-search-production-1.jpg)

||
|:-:|
|**"I tried to restore a snapshot, and now everything's broken."**|

> ❓ **Use Case:** A digital publisher endured a catastrophic outage and attempted to restore from backups. They discovered an index format mismatch only after partial data was lost. Another company saw query performance drop when snapshot compression hogged CPU during peak traffic.

Reviewer 2: It sounds like it was a common problem, so I'm a bit worried at this point.

✅ **Follow Principle of Least Privilege** when configuring access—only give permissions that are absolutely necessary.

For users of Qdrant's managed cloud service, there's an option to configure RBAC directly through the user interface, which automatically creates the role-based access control without requiring manual JWT configuration.

||
|-|
|**Read More:** [**Cloud RBAC Documentation**](https://qdrant.tech/documentation/cloud-rbac/)|

### Remember to Avoid These Common Pitfalls
![vector-search-production](/articles_data/vector-search-production/vector-search-production-8.jpg)

❌ Don't forget to index payload fields—**not doing this will slow your search.**

❌ Don't run without replication—**single-node setups are fragile**.

❌ Don't create a collection per user/customer—**use multitenancy**.

❌ Don't run latency-critical search alongside heavy batch jobs—**separate workloads**.

> ❌ Don't skip quantization—it can **greatly reduce memory and storage footprint**.

Reviewer: Quantization does not reduce storage, because it stores both quantized and non-quantized vectors on disk. It does however speed up searches.

```suggestion
❌ Don't skip quantization—it can **greatly reduce memory footprint and speed up searches**.
```
